# Importing Data

In [37]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [38]:
path = r'C:\Users\Riley\Desktop\Data Any\DD-MM-2022-Instacart Basket Analysis'

In [39]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', '4.3_orders_products', 'products.csv'), index_col = False)

In [40]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [41]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', '4.3_orders_products', 'orders_wrangled.csv'), index_col = False)

In [42]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# 02 Checking for Mixed-Type Data

In [43]:
# Create a data frame
df_test = pd.DataFrame()

In [44]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [45]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [46]:
#Check for mixed-type columns

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [47]:
# Revision
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)
  

product_name


In [48]:
# Fix the mixed-type column
df_test['mix'] = df_test['mix'].astype('str')

# 03 Missing Values

In [49]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [50]:
# Create a subset with the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [51]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [52]:
# New data frame with clean data
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [53]:
df_prods['product_name'].value_counts(dropna = False)

NaN                                                     16
Black House Coffee Roasty Stout Beer                     2
Adore Forever Body Wash                                  2
Gluten Free Organic Peanut Butter & Chocolate Cereal     2
Fiber 4g Gummy Dietary Supplement                        2
                                                        ..
Kosher Dill Mini Pickles                                 1
Chili Rice Scooter                                       1
All Natural Chicken Apple Sausage                        1
Quartered Artichoke Hearts                               1
Fresh Foaming Cleanser                                   1
Name: product_name, Length: 49673, dtype: int64

In [54]:
df_prods.shape

(49693, 5)

In [55]:
df_prods_clean.shape

(49677, 5)

In [56]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [57]:
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

# Duplicates

In [58]:
df_ords.shape

(3421083, 8)

In [59]:
df_prods.shape

(49693, 5)

In [60]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [61]:
df_prods_clean.shape

(49677, 5)

In [62]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [63]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [64]:
# Revision
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [69]:
# Revision
df_prods_clean_no_dups.shape

(49672, 5)

# Questions 2-9

In [65]:
#2 
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Looking at the max value of 99999, which seems suspicious and should be looked into.

In [66]:
#3
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [68]:
df_prods_clean_no_dups.shape

(49672, 5)

#4 
NA no mixed types present

In [ ]:
#5
df_ords.isnull().sum()

In [ ]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [ ]:
df_ords_nan

The days_since_prior_order column has many missing values, which may be due to the lack of a second order. This would create a lot of missing values since these people couldn't have a days since prior order stat.

#6
In order to fix these missing value you would need to make a new column to keep track of first time orders, and then manipulate them some other way. Deleting the missing values as is would wreak havoc on the data, as it would delete important entries.

In [ ]:
df_ords_clean = df_ords

In [ ]:
df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [ ]:
df_ords_clean.head()

In [ ]:
#7 
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [ ]:
df_ords_dups

No duplicates since the output is empty

In [ ]:
#8
df_ords_dups.shape

No duplicates present

In [ ]:
#9
df_ords_clean.to_csv(os.path.join(path, '02 Data','4.3_orders_products', 'orders_cleaned.csv'))

In [ ]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','4.3_orders_products', 'products_cleaned.csv'))